src/net/NetworkServer.cpp

```diff
@@ -0,0 +1,149 @@
+#include "NetworkServer.h"
+
+TcpConnection::pointer TcpConnection::create(boost::asio::io_service& io_service) {
+  return pointer(new TcpConnection(io_service));
+}
+
+tcp::socket& TcpConnection::get_socket() {
+  return socket;
+}
+
+void TcpConnection::start() {
+  socket.async_read_some(boost::asio::buffer(rcvbuf),
+      boost::bind(&TcpConnection::handle_read, shared_from_this(),
+        boost::asio::placeholders::error,
+        boost::asio::placeholders::bytes_transferred));
+}
+
+bool TcpConnection::send(std::string message) {
+  if (!socket.is_open()) {
+    return false;
+  }
+  
+  boost::system::error_code error;
+  boost::asio::write(socket, boost::asio::buffer(message), error);
+  
+  // If we get these errors, it's likely a clean disconnect.
+  if ((error == boost::asio::error::eof) ||
+      (error == boost::asio::error::connection_reset) ||
+      (error == boost::asio::error::broken_pipe)) {
+    std::cerr << "[send] client disconnected.\n";
+    return false;
+  } else if (error) {
+    std::cerr << "[send] some other error: " << error << "\n";
+    return false;
+  }
+  
+  return true;
+}
+
+bool TcpConnection::has_messages() {
+  return !message_queue.empty();
+}
+
+std::string TcpConnection::pop_message() {
+  if (!has_messages()) 
+    throw std::logic_error("No messages"); 
+  return message_queue.pop(); 
+}
+
+TcpConnection::TcpConnection(boost::asio::io_service& io_service)
```

Comment:
Why is this constructor just in the middle of all the other definitions :angery: